### PR TITLE
Refacto : add a Transport class

### PR DESCRIPTION
**Pull request by rxlabz**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring

## :scroll: Description
send the event from a dedicated class

## :bulb: Motivation and Context

https://develop.sentry.dev/sdk/unified-api/

## :green_heart: How did you test it?

updated existing tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/browser_client.dart [hunk 1]
```diff
@@ -41,19 +42,19 @@ class SentryBrowserClient extends SentryClient {
...
       : super.base(
           options,
-          origin: origin,
-          sdk: Sdk(name: browserSdkName, version: sdkVersion),
-          platform: platform,
+          transport: Transport(
+            dsn: options.dsn,
+            httpClient: options.httpClient,
+            clock: options.clock,
+            compressPayload: false,
```
Comment (marandaneto) [inline dart/lib/src/browser_client.dart]:
```suggestion
            compressPayload: options.compressPayload,
```

File: dart/lib/src/browser_client.dart [hunk 2]
```diff
@@ -41,19 +42,19 @@ class SentryBrowserClient extends SentryClient {
...
+            sdk: Sdk(name: browserSdkName, version: sdkVersion),
```
Comment (marandaneto) [inline dart/lib/src/browser_client.dart]: `options.sdkVersion` already exists, but its probably null, we'd need to find a way to instantiate it based on the platform (dart, flutter or web), but ideally, we just read this field from the options here

File: dart/lib/src/browser_client.dart [hunk 3]
```diff
@@ -41,19 +42,19 @@ class SentryBrowserClient extends SentryClient {
...
+            headersBuilder: SentryClient.buildHeaders,
```
Comment (marandaneto) [inline dart/lib/src/browser_client.dart]: this could be done inside of the Transport layer I guess, why does the client need to know about the headers?

File: dart/lib/src/browser_client.dart [hunk 4]
```diff
@@ -41,19 +42,19 @@ class SentryBrowserClient extends SentryClient {
...
+            platform: platform,
```
Comment (marandaneto) [inline dart/lib/src/browser_client.dart]: wondering if this is the same as `options.sentryClientName`, could you check it?

File: dart/lib/src/client.dart [hunk 5]
```diff
@@ -22,36 +21,14 @@ abstract class SentryClient {
...
-  /// The Sentry.io secret key for the project.
   @visibleForTesting
-  // ignore: invalid_use_of_visible_for_testing_member
-  String get secretKey => _dsn.secretKey;
-
-  /// The ID issued by Sentry.io to your project.
-  ///
-  /// Attached to the event payload.
-  String get projectId => _dsn.projectId;
+  final Transport transport;
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: ideally the transport is read from the `options. transport ` so we don't need to `visibleForTesting`, and we only instantiate a new `transport` here if the current `optins.transport` is a NoOp.
That said, we need a NoOpTransport: https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/transport/NoOpTransport.java

let's ignore the `retryAfter` for now (https://github.com/getsentry/sentry-dart/projects/1#card-47870442)

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: as the payload compression is not available in browser should we let the choice to the user ?

File: dart/lib/src/client.dart [hunk 6]
```diff
@@ -188,15 +109,63 @@ abstract class SentryClient {
...
 
-  @override
-  String toString() => '$SentryClient("$postUri")';
+  SentryEvent _processEvent(
+    SentryEvent event, {
+    dynamic hint,
+    List<EventProcessor> eventProcessors,
+  }) {
+    for (final processor in eventProcessors) {
+      event = processor(event, hint);
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: if one of the processors returned null, means that this is event is dropped, so we should bail out and log it.
eg https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryClient.java#L155-L184

Comment (marandaneto) [reply dart/lib/src/client.dart]: I can help with those things on a different PR though

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: ok, I missed that, if the browser client doesn't support that, we should not give an option for sure, but I feel this should be documented in the `compressPayload` field declaration then.

File: dart/lib/src/client.dart [hunk 7]
```diff
@@ -188,15 +109,63 @@ abstract class SentryClient {
...
+      // Merge the scope extra.
+      if (event.extra == null) {
+        event = event.copyWith(extra: scope.extra);
+      }
+
+      // Merge the scope level.
+      if (event.level == null) {
+        event = event.copyWith(level: scope.level);
+      }
+    }
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: for some fields, we actually append items from the Scope to the event, see:
https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryClient.java#L281-L335

File: dart/lib/src/client.dart [hunk 8]
```diff
@@ -188,15 +109,63 @@ abstract class SentryClient {
...
+    return event;
+  }
+
+  @override
+  String toString() => '$SentryClient("${options.dsn}")';
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: mm not against it, but I don't see how this is useful :)

File: dart/lib/src/transport/transport.dart [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+  final bool compressPayload;
+
+  /// Used by sentry to differentiate browser from io environment
+  final String platform;
+
+  final ClockProvider _clock;
+
+  /// Use for browser stacktrace
+  final String origin;
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: most of those fields if not all are already in the options, should it not just take options and read from there?

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: is `compressPayload` really an option here ? 🤔
Could you choose to send a non compressed payload if you're on a platform which supports `gzip` ?

File: dart/lib/src/transport/transport.dart [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    String apiPath;
+    if (pathLength > 1) {
+      // some paths would present before the projectID in the dsnUri
+      apiPath =
+          (dsnUri.pathSegments.sublist(0, pathLength - 1) + ['api']).join('/');
+    } else {
+      apiPath = 'api';
+    }
+    return '${dsnUri.scheme}://${dsnUri.host}$port/$apiPath/$projectId/store/';
+  }
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: does it make sense to move this to the `Dsn` class like https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Dsn.java

ideally, Dsn would offer atributes like `dsn.projectId`, etc...

File: dart/lib/src/transport/transport.dart [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+  Future<SentryId> send(Map<String, dynamic> data) async {
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: our Transport still takes a typed object, `SentryEvent` in the case, and only here in the transport we'd convert to a json-like object
https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/transport/ITransport.java#L11

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: we could use a conditional import in `transport.dart` to import the bodyEncoder strategy ? and let it be a transport "hidden" implementation details ?

File: dart/lib/src/transport/transport.dart [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    final now = _clock();
+    var authHeader = 'Sentry sentry_version=6, sentry_client=$clientId, '
+        'sentry_timestamp=${now.millisecondsSinceEpoch}, sentry_key=$publicKey';
+    if (secretKey != null) {
+      authHeader += ', sentry_secret=$secretKey';
+    }
+
+    mergeAttributes(_getContext(now), into: data);
+
+    final headers = headersBuilder(authHeader);
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: on Java we have an abstraction for this, so we only do it once, and not on every call to `send(...)`, see: https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/CredentialsSettingConfigurator.java
could we create such abstraction or at least build it only once on the transport? I think the only moving part here is the `sentry_timestamp`

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: I forgot that it's already a conditional import in `transport.dart`. the question is : should we hide `compressPayload ` this from the public API and set it based on the current platform ?

File: dart/lib/src/transport/transport.dart [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+          (dsnUri.pathSegments.sublist(0, pathLength - 1) + ['api']).join('/');
+    } else {
+      apiPath = 'api';
+    }
+    return '${dsnUri.scheme}://${dsnUri.host}$port/$apiPath/$projectId/store/';
+  }
+
+  Future<SentryId> send(Map<String, dynamic> data) async {
+    final now = _clock();
+    var authHeader = 'Sentry sentry_version=6, sentry_client=$clientId, '
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: latest sentry_version is 7, but we'd need to check if there are differences, I've created a separated issue for that
https://github.com/getsentry/sentry-dart/projects/1#card-47871908

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: not really, for flutter, they still can opt in-out of the compressPayload, lets not make this limitation.
lets keep uncompressed for web, and later we fix this too https://github.com/getsentry/sentry-dart/projects/1#card-47870976

File: dart/lib/src/browser_client.dart (hunk 3 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: updated  : now it uses a conditional platform in transport.

File: dart/lib/src/transport/transport.dart (hunk 9 above)
Comment (rxlabz) [reply dart/lib/src/transport/transport.dart]: Updated : the transport is now instantiated from a SentryOptions

File: dart/lib/src/transport/transport.dart (hunk 10 above)
Comment (rxlabz) [reply dart/lib/src/transport/transport.dart]: updated : all this logic is now in the `Dsn`class

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: it's not an option in Java, is it ? https://github.com/getsentry/sentry-java/blob/67669e1c6d727af618424b7542e35db5c7797c54/sentry/src/main/java/io/sentry/transport/HttpTransport.java#L254-L256 
As Flutter web needs the same behavior than other Dart web projects, I don't really understand why the user could set it ?

Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: its not, but it should, we have this in our Java backlog

File: dart/lib/src/client.dart (hunk 6 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅  fixed

File: dart/lib/src/client.dart (hunk 7 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅  fixed

File: dart/lib/src/client.dart (hunk 8 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: I don't know :) until know the postUri was returned... what kind of infos should we return here ? Do we need it ?

Comment (marandaneto) [reply dart/lib/src/client.dart]: I dont think we need to overwrite the toString for this tbh, makes more sense for data classes only IMO

File: dart/lib/src/transport/transport.dart (hunk 11 above)
Comment (rxlabz) [reply dart/lib/src/transport/transport.dart]: ✅  fixed

File: dart/lib/src/transport/transport.dart (hunk 12 above)
Comment (rxlabz) [reply dart/lib/src/transport/transport.dart]: ✅  I added a CredentialBuilder

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (bruno-garcia) [reply dart/lib/src/browser_client.dart]: Could just be a no-op for Web (Dart or Flutter), since it doesn't apply.
Conditionally having the option will make it harder for users to consume the options since they'll need to conditionally set a value depending on the platform, feels like lots of friction

File: dart/lib/src/browser_client.dart (hunk 2 above)
Comment (bruno-garcia) [reply dart/lib/src/browser_client.dart]: on a second thought, shouldn't this be just either: `sentry.dart` or `sentry.flutter`? Given that this is what the user installed at the end?

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: a condition for importing dart or web http packages already exist, so the only difference here is if they use `web`, and set  `compressPayload=true` won't have any effect till we implement it

File: dart/lib/src/browser_client.dart (hunk 2 above)
Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: if we need only the identifier of the client, yes, only a String would be fine.

File: dart/lib/src/client.dart [hunk 14]
```diff
@@ -188,48 +86,69 @@ abstract class SentryClient {
...
+
+      // Merge the scope extra.
+      event = event.copyWith(
+          extra: scope.extra.map((key, value) => MapEntry(key, value))
+            ..addAll(event.extra ?? {}));
+
+      // Merge the scope level.
+      if (event.level == null) {
+        event = event.copyWith(level: scope.level);
+      }
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: Scope level has precedence here:
https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryClient.java#L327-L330

File: dart/lib/src/transport/transport.dart [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+      dsn: Dsn.parse(options.dsn),
+      clientId: sdk.identifier,
+    );
+  }
+
+  Future<SentryId> send(SentryEvent event) async {
+    final now = _options.clock();
+
+    var authHeader = _credentialBuilder.build(now.millisecondsSinceEpoch);
+    final headers = buildHeaders(authHeader, sdk: sdk);
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: this could be assigned to a private global field on the ctor body, so we don't need to do on every event

File: dart/lib/src/transport/transport.dart [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    };
+
+    if (_options.environmentAttributes != null) {
+      mergeAttributes(_options.environmentAttributes.toJson(), into: data);
+    }
+
+    mergeAttributes(
+      event.toJson(origin: origin),
+      into: data,
+    );
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: this should be the SentryClient's job if I got it right, the transport is only responsible to send it, the event should be fully built and not modified at this point.

Comment (marandaneto) [reply dart/lib/src/transport/transport.dart]: I missed the selection for the `_getContext` method here but it's pretty much the same

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (bruno-garcia) [reply dart/lib/src/browser_client.dart]: I'm aware, but there's no gzip on web, the browser will gzip or not and we don't control it

Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: if this is not controlled on the client, totally in for NoOp, I thought this was a limitation because we didn't implement it

File: dart/lib/src/noop_client.dart [hunk 17]
```diff
@@ -21,17 +22,7 @@ class NoOpSentryClient implements SentryClient {
...
-  @override
-  List<int> bodyEncoder(
-    Map<String, dynamic> data,
-    Map<String, String> headers,
-  ) =>
-      [];
-
-  @override
-  Map<String, String> buildHeaders(String authHeader) => {};
+  Transport transport;
```
Comment (marandaneto) [inline dart/lib/src/noop_client.dart]: why is this necessary? one should not care about the `transport` when implementing `SentryClient`, transport comes from `options`

File: dart/lib/src/client.dart [hunk 18]
```diff
@@ -188,48 +86,69 @@ abstract class SentryClient {
...
+        );
+      }
+      if (event == null) {
+        options.logger(SentryLevel.debug, 'Event was dropped by a processor');
+        break;
+      }
     }
-
-    return headers;
+    return event;
```
Comment (marandaneto) [inline dart/lib/src/client.dart:120]: still missing the execution of the beforeSendCallback ->
https://github.com/getsentry/sentry-java/blob/5ebd49cbfd17eb66175e9247db6d27e4df84a56b/sentry/src/main/java/io/sentry/SentryClient.java#L97-L103

File: dart/lib/src/client.dart (hunk 14 above)
Comment (marandaneto) [reply dart/lib/src/client.dart]: also, missing Contexts, created a card for that https://github.com/getsentry/sentry-dart/projects/1#card-47942441
it doesn't need to be on this PR

File: dart/lib/src/browser_client.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: we will address this on other PR https://github.com/getsentry/sentry-dart/projects/1#card-47942558

File: dart/lib/src/client.dart [hunk 19]
```diff
@@ -19,38 +17,13 @@ abstract class SentryClient {
...
-    this.options, {
-    String platform,
-    this.origin,
-    Sdk sdk,
-  })  : _dsn = Dsn.parse(options.dsn),
-        _platform = platform ?? sdkPlatform,
-        sdk = sdk ?? Sdk(name: sdkName, version: sdkVersion);
-
-  final Dsn _dsn;
+  SentryClient.base(this.options, {@required this.transport});
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: `options` already have a `transport` field, lets take from the `options` instead of passing as named param.

File: dart/lib/src/client.dart (hunk 5 above)
Comment (marandaneto) [reply dart/lib/src/client.dart]: we can remove this, even for testing we can set a mock/fake on the options

File: dart/lib/src/client.dart [hunk 20]
```diff
@@ -64,88 +37,22 @@ abstract class SentryClient {
...
-      mergeAttributes(<String, dynamic>{'user': userContext.toJson()},
-          into: data);
-    }
-
-    mergeAttributes(
-      event.toJson(
-        origin: origin,
-      ),
-      into: data,
+    event = event.copyWith(
```
Comment (marandaneto) [inline dart/lib/src/client.dart:48]: projectId is not really necessaary, its part of the DSN URL
https://develop.sentry.dev/sdk/event-payloads/

File: dart/lib/src/io_client.dart [hunk 21]
```diff
@@ -2,60 +2,29 @@
...
-    if (options.compressPayload) {
-      headers['Content-Encoding'] = 'gzip';
-      body = gzip.encode(body);
-    }
-    return body;
-  }
+  SentryIOClient._(SentryOptions options, {@required String platform})
+      : super.base(
+          options,
+          transport: Transport(options: options, sdk: sdk, platform: platform),
```
Comment (marandaneto) [inline dart/lib/src/io_client.dart]: https://github.com/getsentry/sentry-java/blob/5ebd49cbfd17eb66175e9247db6d27e4df84a56b/sentry/src/main/java/io/sentry/SentryClient.java#L43-L47

Comment (marandaneto) [reply dart/lib/src/io_client.dart]: we set the `transport` to the `options` and then pass the `options` along, so `transport` wont be null

Comment (marandaneto) [reply dart/lib/src/io_client.dart]: just to remember: we might be able to identify if its web or not on runtime via https://api.flutter.dev/flutter/foundation/kIsWeb-constant.html
so transport wont even need a platform param here, it could be already on sentry options ctor for example

File: dart/lib/src/transport/header_builder.dart [hunk 22]
```diff
@@ -0,0 +1,14 @@
+import '../protocol.dart';
+
+Map<String, String> buildHeaders(/*String authHeader, */ {Sdk sdk}) {
```
Comment (marandaneto) [inline dart/lib/src/transport/header_builder.dart]: it could be a String identifier, instead of the Sdk instance

File: dart/lib/src/transport/header_builder.dart [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+  final headers = {
+    'Content-Type': 'application/json',
+    // NOTE(lejard_h) overriding user agent on VM and Flutter not sure why
+    // for web it use browser user agent
+    'User-Agent': sdk.identifier
+  };
+
+  headers['User-Agent'] = sdk.identifier;
```
Comment (marandaneto) [inline dart/lib/src/transport/header_builder.dart]: its duplicated

File: dart/lib/src/transport/transport.dart [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+  @visibleForTesting
+  final Dsn dsn;
+
+  /// Use for browser stacktrace
+  final String origin;
+
+  /// Used by sentry to differentiate browser from io environment
+  final String platform;
+
+  final Sdk sdk;
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: if they are already in the options, they dont need to be public global fields

File: dart/lib/src/transport/transport.dart [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  Map<String, dynamic> _getEventData(
+    SentryEvent event, {
+    DateTime timeStamp,
+  }) {
+    final data = event.toJson(origin: origin);
+
+    // TODO add this attributes to event in client
+    if (_options.environmentAttributes != null) {
```
Comment (marandaneto) [inline dart/lib/src/transport/transport.dart]: yep, `environmentAttributes` should be even gone, all the values read from `environmentAttributes` should be added on the options, if not there yet

File: dart/lib/src/browser_client.dart (hunk 2 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: ✅ fixed

File: dart/lib/src/browser_client.dart (hunk 4 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: `platform` is `'dart'` or `'javascript'` , according to the docs ``options.sentryClientName`` is `'sentry.{language}.{platform}/{version}'`

Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: we could do this then?

```
factory SentryIOClient(SentryOptions options) {
    options.platform = sdkPlatform;
    SentryIOClient._(options);
}
```

same for the WebClient, so we move this flag to the options and the `transport` don't need to take it as a param? we do it automatically.

File: dart/lib/src/client.dart (hunk 20 above)
Comment (rxlabz) [reply dart/lib/src/client.dart:48]: ✅ fixed

File: dart/lib/src/transport/header_builder.dart (hunk 23 above)
Comment (rxlabz) [reply dart/lib/src/transport/header_builder.dart]: ✅

File: dart/lib/src/client.dart [hunk 26]
```diff
@@ -64,88 +37,19 @@ abstract class SentryClient {
   /// * https://docs.sentry.io/learn/context/#capturing-the-user
   User userContext;
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: this should be part of the SentryClient only as well, it belongs to the Scope.user

File: dart/lib/src/client.dart [hunk 27]
```diff
@@ -64,88 +37,19 @@ abstract class SentryClient {
...
-      postUri,
-      headers: headers,
-      body: body,
-    );
+    event = _applyScope(event: event, scope: scope);
 
-    if (response.statusCode != 200) {
-      return SentryId.empty();
-    }
+    event = event.copyWith(platform: transport.platform);
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: why do we need to read the `platform` from the `transport`? its already defining on the SentryEvent ctor based on `isWeb`.
I believe `transport.platform` could be done altogether

File: dart/lib/src/client.dart [hunk 28]
```diff
@@ -17,13 +17,19 @@ abstract class SentryClient {
...
+      );
+    }
+  }
 
   @protected
-  SentryOptions options;
-
   @visibleForTesting
-  final Transport transport;
+  SentryOptions options;
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: it could be private I believe

File: dart/lib/src/client.dart [hunk 29]
```diff
@@ -19,38 +18,19 @@ abstract class SentryClient {
...
-  /// The Sentry.io public key for the project.
-  @visibleForTesting
-  // ignore: invalid_use_of_visible_for_testing_member
-  String get publicKey => _dsn.publicKey;
+  SentryClient.base(this.options, {String origin}) {
+    _random = options.sampleRate == null ? null : Random();
+    if (options.transport is NoOpTransport) {
+      options.transport = Transport(
+        options: options,
+        sdkIdentifier: '${sdkName}/${sdkVersion}',
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: could this be replaced by `options.sdkVersion.identifier`? it looks duplicated now

File: dart/lib/src/browser_client.dart (hunk 4 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: ✅  platform is now defined at runtime

File: dart/lib/src/protocol/sentry_event.dart [hunk 30]
```diff
@@ -31,10 +32,10 @@ class SentryEvent {
     this.userContext,
     this.contexts,
     this.breadcrumbs,
-    this.sdk,
   })  : eventId = eventId ?? SentryId.newId(),
-        platform = platform ?? sdkPlatform,
-        timestamp = timestamp ?? getUtcDateTime();
+        platform = platform ?? (isWeb ? browserPlatform : sdkPlatform),
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]: if this is going to be set here now, why do we need this on the client/transport? I believe this  should be done on the Client, like Java does (via MainEventProcessor)

File: dart/lib/src/client.dart (hunk 5 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅

File: dart/lib/src/sentry_options.dart [hunk 31]
```diff
@@ -139,8 +148,9 @@ class SentryOptions {
     this.environmentAttributes,
     this.compressPayload,
     this.httpClient,
+    Transport transport,
     ClockProvider clock = getUtcDateTime,
-  }) {
+  }) : _transport = transport ?? NoOpTransport() {
```
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]: why do we have this here? its already `NoOpTransport` by default, we expose a setter already

File: dart/lib/src/transport/noop_transport.dart [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+
+class NoOpTransport implements Transport {
+  @override
+  Dsn get dsn => null;
+
+  @override
+  String get origin => null;
+
+  @override
+  String get sdkIdentifier => null;
```
Comment (marandaneto) [inline dart/lib/src/transport/noop_transport.dart]: this should not be part of the public API, they should only care about the `send` method

File: dart/lib/src/version.dart [hunk 33]
```diff
@@ -8,19 +8,28 @@
 /// This library contains Sentry.io SDK constants used by this package.
 library version;
 
+import 'utils.dart';
+
 /// The SDK version reported to Sentry.io in the submitted events.
 const String sdkVersion = '4.0.0';
 
+String get sdkName => isWeb ? browserSdkName : ioSdkName;
```
Comment (marandaneto) [inline dart/lib/src/version.dart]: we can make those const private now, if what matters is the `sdkName`

File: dart/lib/src/client.dart (hunk 27 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅

File: dart/lib/src/client.dart (hunk 28 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅ fixed

File: dart/lib/src/client.dart (hunk 29 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅ my bad, I thought `sdkVersion` was just a String, I renamed it `sdk`, and define it in the ioClient & BrowserClient if null

File: dart/lib/src/sentry_options.dart (hunk 31 above)
Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: we can't access the setter from the constructor, so if we don't want a null transport we needs this.

Comment (marandaneto) [reply dart/lib/src/sentry_options.dart]: but the field declaration already has NoOpTransport as a value, it should not belong to the ctor, a setter wont be able to set it to null either, because on the setter we need to check if its null and not allow to set it

File: dart/lib/src/transport/transport.dart (hunk 24 above)
Comment (rxlabz) [reply dart/lib/src/transport/transport.dart]: for now I just let the `dsn` which is used for a lot of tests ( marked as @visibleForTestings )

File: dart/lib/src/transport/noop_transport.dart (hunk 32 above)
Comment (rxlabz) [reply dart/lib/src/transport/noop_transport.dart]: ✅  I just let `dsn` for now as it is used in a lot of tests

File: dart/lib/src/version.dart (hunk 33 above)
Comment (rxlabz) [reply dart/lib/src/version.dart]: ✅

File: dart/lib/src/sentry_options.dart (hunk 31 above)
Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: we can't write this : 
```dart
SentryOptions({
    this.dsn,
    this.environmentAttributes,
    this.compressPayload,
    this.httpClient,
    this.transport,
    ClockProvider clock = getUtcDateTime,
  })
```
cause this.transport refers to setter and it can't be accessed from here

we can remove the default value in the declaration of the field, but I think `: _transport = transport ?? NoOpTransport()` is needed

Comment (marandaneto) [reply dart/lib/src/sentry_options.dart]: But we dont need a transport in the ctor, as i said, we dont need any ctor params, options is a mutable class and all fields have setters, you could even remove this ctor altogether

Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: ah sorry. in french we say : "I understand fast, but you have to explain me long time" :D

Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: ✅ removed !

File: dart/lib/src/transport/transport.dart (hunk 25 above)
Comment (rxlabz) [reply dart/lib/src/transport/transport.dart]: ✅  removed

File: dart/lib/src/protocol/sentry_event.dart (hunk 30 above)
Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: ✅ I moved it to the client.

File: dart/lib/src/client.dart (hunk 26 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅

File: dart/lib/src/client.dart (hunk 8 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: ✅

**Review by marandaneto — APPROVED** [review]
nice, thanks for doing this, let's iterate from feature/unified-api
